### ISSUE TITLE
[await-thenable] react-dom/test-utils "act" function not playing nicely

### ISSUE THREAD

**Issue by rbutera** [typescript-eslint/typescript-eslint#1913]
**Repro**

### package.json
```JSON
{
   "dependencies": {
    "react-dom": "^16.13.1",
    "react": "^16.13.1",
  },
  "devDependencies": {
    "@types/react-dom": "^16.9.6",
    "@types/react": "^16.9.34",
    "@typescript-eslint/eslint-plugin": "^2.28.0",
    "@typescript-eslint/parser": "^2.28.0",
	"@testing-library/react": "^10.0.2",
    "eslint": "^6.8.0",
    "typescript": "^3.9.0-beta"
  }
}
```

### example test

```TS
import React from "react"
import {render, fireEvent} from '@testing-library/react'
import {act} from 'react-dom/test-utils'


const Form: React.FunctionComponent = () => (<form data-testid="form"><input type="text" name="foo" value="bar"/></form>)

const {getByTestId} = render(<Form>)
const form = getByTestId("form")

const result = await act(async () => { // THIS IS WHERE ERROR IS TRIGGERED
  await fireEvent.submit(form)
})
```

### `act`'s type definition

```TS
// ./node_modules/@types/react-dom/test-utils/index.d.ts
/**
 * Wrap any code rendering and triggering updates to your components into `act()` calls.
 *
 * Ensures that the behavior in your tests matches what happens in the browser
 * more closely by executing pending `useEffect`s before returning. This also
 * reduces the amount of re-renders done.
 *
 * @param callback A synchronous, void callback that will execute as a single, complete React commit.
 *
 * @see https://reactjs.org/blog/2019/02/06/react-v16.8.0.html#testing-hooks
 */
// the "void | undefined" is here to forbid any sneaky "Promise" returns.
export function act(callback: () => void | undefined): void;
// the "void | undefined" is here to forbid any sneaky return values
// tslint:disable-next-line: void-return
export function act(callback: () => Promise<void | undefined>): Promise<undefined>;
```

**Expected Result**

no error as `act` is thenable (see act's type definition above)

**Actual Result**

`Unexpected `await` of a non-Promise (non-"Thenable") value.`

**Additional Info**

`act` has been thenable since react 16.9. I have no idea what is going on here.

**Comment by bradzacher** [comment]
Could you please provide a repro repo?

It's hard to investigate this as is, as there are a number of packages and versions that may need to align.

**Comment by rbutera** [comment]
@bradzacher
thanks for the quick response!

I've created a repro repo:

https://github.com/rbutera/typescript-eslint-1913-await-thenable

**Comment by bradzacher** [comment]
thanks for the repro. from that it was pretty trivial to find the problem.
the problem is that typescript is picking the `void` signature instead of the `Promise<undefined>` signature.

Why? because despite the fact that their comment says `the "void | undefined" is here to forbid any sneaky "Promise" returns.`, it doesn't actually work - typescript can still assign `Promise` to `void`, so it picks that signature.

This is pretty easy to fix, but it'll have to be from the types side.
Typescript inspects the signatures in source order.
So simply switching the order of the signatures is enough to fix this issue!
Because I've got the repo locally, I'll submit a PR to fix the types.

Note - another way to fix this for yourself is to use `strictNullChecks`.
